File: old/tags/2.1/java/edu/nps/moves/dis/GridAxisRecordRepresentation1.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;
import edu.nps.moves.jaxb.dis.*;

/**
 * 5.2.44: Grid data record, representation 1
 *
 * Copyright (c) 2007, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 */
public class GridAxisRecordRepresentation1 extends GridAxisRecord implements Serializable
{
   /** constant scale factor */
   protected float  fieldScale;

   /** constant offset used to scale grid data */
   protected float  fieldOffset;

   /** Number of data values */
   protected int  numberOfValues;

   /** variable length list of data parameters @@@this is wrong--need padding as well */
   protected List dataValues = new ArrayList(); 

/** Constructor */
 public GridAxisRecordRepresentation1()
 {
 }

/** 
 * Constructor--takes a parallel jaxb object and returns an open-dis object 
 * 1.4_sed_bait_start */
 public GridAxisRecordRepresentation1(edu.nps.moves.jaxb.dis.GridAxisRecordRepresentation1 x)
 {
     super(x); // Call superclass constructor

     this.fieldScale = x.getFieldScale();
     this.fieldOffset = x.getFieldOffset();
     this.numberOfValues = x.getNumberOfValues();
     this.dataValues = new ArrayList();
     for(int idx = 0; idx < x.getDataValues().size(); idx++)
     {
        this.dataValues.add( new edu.nps.moves.dis.TwoByteChunk((edu.nps.moves.jaxb.dis.TwoByteChunk) x.getDataValues().get(idx)));
     }
 }
/* 1.4_sed_bait_end */


/**
 * returns a jaxb object intialized from this object, given an empty jaxb object
 * 1.4_sed_bait_start **/
 public edu.nps.moves.jaxb.dis.GridAxisRecordRepresentation1 initializeJaxbObject(edu.nps.moves.jaxb.dis.GridAxisRecordRepresentation1 x)
 {
     super.initializeJaxbObject(x); // Call superclass initializer

     ObjectFactory factory = new ObjectFactory();

     x.setFieldScale( this.getFieldScale() );
     x.setFieldOffset( this.getFieldOffset() );
     x.setNumberOfValues( this.getNumberOfValues() );

     List dataValues_1 = x.getDataValues();
     for(int idx = 0; idx < dataValues.size(); idx++)
     {
         TwoByteChunk a = (edu.nps.moves.dis.TwoByteChunk)dataValues.get(idx);
         dataValues_1.add(a.initializeJaxbObject(factory.createTwoByteChunk()));
     }
   return x;
 }
/* 1.4_sed_bait_end */


public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = super.getMarshalledSize();
   marshalSize = marshalSize + 4;  // fieldScale
   marshalSize = marshalSize + 4;  // fieldOffset
   marshalSize = marshalSize + 2;  // numberOfValues
   for(int idx=0; idx < dataValues.size(); idx++)
   {
        TwoByteChunk listElement = (TwoByteChunk)dataValues.get(idx);
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setFieldScale(float pFieldScale)
{ fieldScale = pFieldScale;
}

public float getFieldScale()
{ return fieldScale; 
}

public void setFieldOffset(float pFieldOffset)
{ fieldOffset = pFieldOffset;
}

public float getFieldOffset()
{ return fieldOffset; 
}

public int getNumberOfValues()
{ return (int)dataValues.size();
}

/** Note that setting this value will ot change the marshalled value. The list whose length this describes is used for that purpose.
 * The getnumberOfValues method will also be based on the actual list length rather than this value. 
 * The method is simply here for java bean completeness.
 */
public void setNumberOfValues(int pNumberOfValues)
{ numberOfValues = pNumberOfValues;
}

public void setDataValues(List pDataValues)
{ dataValues = pDataValues;
}

public List getDataValues()
{ return dataValues; }


public void marshal(DataOutputStream dos)
{
    super.marshal(dos);
    try 
    {
       dos.writeFloat( (float)fieldScale);
       dos.writeFloat( (float)fieldOffset);
       dos.writeShort( (short)dataValues.size());

       for(int idx = 0; idx < dataValues.size(); idx++)
       {
            TwoByteChunk aTwoByteChunk = (TwoByteChunk)dataValues.get(idx);
            aTwoByteChunk.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      System.out.println(e);}
    } // end of marshal method

public void unmarshal(DataInputStream dis)
{
    super.unmarshal(dis);

    try 
    {
       fieldScale = dis.readFloat();
       fieldOffset = dis.readFloat();
       numberOfValues = dis.readShort();
        for(int idx = 0; idx < numberOfValues; idx++)
        {
           TwoByteChunk anX = new TwoByteChunk();
            anX.unmarshal(dis);
            dataValues.add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      System.out.println(e); 
    }
 } // end of unmarshal method 


 /**
  * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
  */
 public boolean equals(GridAxisRecordRepresentation1 rhs)
 {
     boolean ivarsEqual = true;

    if(rhs.getClass() != this.getClass())
        return false;

     if( ! (fieldScale == rhs.fieldScale)) ivarsEqual = false;
     if( ! (fieldOffset == rhs.fieldOffset)) ivarsEqual = false;
     if( ! (numberOfValues == rhs.numberOfValues)) ivarsEqual = false;

     for(int idx = 0; idx < dataValues.size(); idx++)
     {
        TwoByteChunk x = (TwoByteChunk)dataValues.get(idx);
        if( ! ( dataValues.get(idx).equals(rhs.dataValues.get(idx)))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class

File: old/tags/2.2/cpp/DIS/GridAxisRecordRepresentation1.cpp
#include <DIS/GridAxisRecordRepresentation1.h> 

using namespace DIS;


GridAxisRecordRepresentation1::GridAxisRecordRepresentation1() : GridAxisRecord(),
   _fieldScale(0.0), 
   _fieldOffset(0.0), 
   _numberOfValues(0)
{
}

GridAxisRecordRepresentation1::~GridAxisRecordRepresentation1()
{
    _dataValues.clear();
}

float GridAxisRecordRepresentation1::getFieldScale() const
{
    return _fieldScale;
}

void GridAxisRecordRepresentation1::setFieldScale(float pX)
{
    _fieldScale = pX;
}

float GridAxisRecordRepresentation1::getFieldOffset() const
{
    return _fieldOffset;
}

void GridAxisRecordRepresentation1::setFieldOffset(float pX)
{
    _fieldOffset = pX;
}

unsigned short GridAxisRecordRepresentation1::getNumberOfValues() const
{
   return _dataValues.size();
}

std::vector<TwoByteChunk>& GridAxisRecordRepresentation1::getDataValues() 
{
    return _dataValues;
}

const std::vector<TwoByteChunk>& GridAxisRecordRepresentation1::getDataValues() const
{
    return _dataValues;
}

void GridAxisRecordRepresentation1::setDataValues(const std::vector<TwoByteChunk>& pX)
{
     _dataValues = pX;
}

void GridAxisRecordRepresentation1::marshal(DataStream& dataStream) const
{
    GridAxisRecord::marshal(dataStream); // Marshal information in superclass first
    dataStream << _fieldScale;
    dataStream << _fieldOffset;
    dataStream << ( unsigned short )_dataValues.size();

     for(size_t idx = 0; idx < _dataValues.size(); idx++)
     {
        TwoByteChunk x = _dataValues[idx];
        x.marshal(dataStream);
     }

}

void GridAxisRecordRepresentation1::unmarshal(DataStream& dataStream)
{
    GridAxisRecord::unmarshal(dataStream); // unmarshal information in superclass first
    dataStream >> _fieldScale;
    dataStream >> _fieldOffset;
    dataStream >> _numberOfValues;

     _dataValues.clear();
     for(size_t idx = 0; idx < _numberOfValues; idx++)
     {
        TwoByteChunk x;
        x.unmarshal(dataStream);
        _dataValues.push_back(x);
     }
}


bool GridAxisRecordRepresentation1::operator ==(const GridAxisRecordRepresentation1& rhs) const
 {
     bool ivarsEqual = true;

     ivarsEqual = GridAxisRecord::operator==(rhs);

     if( ! (_fieldScale == rhs._fieldScale) ) ivarsEqual = false;
     if( ! (_fieldOffset == rhs._fieldOffset) ) ivarsEqual = false;

     for(size_t idx = 0; idx < _dataValues.size(); idx++)
     {
        if( ! ( _dataValues[idx] == rhs._dataValues[idx]) ) ivarsEqual = false;
     }


    return ivarsEqual;
 }

int GridAxisRecordRepresentation1::getMarshalledSize() const
{
   int marshalSize = 0;

   marshalSize = GridAxisRecord::getMarshalledSize();
   marshalSize = marshalSize + 4;  // _fieldScale
   marshalSize = marshalSize + 4;  // _fieldOffset
   marshalSize = marshalSize + 2;  // _numberOfValues

   for(int idx=0; idx < _dataValues.size(); idx++)
   {
        TwoByteChunk listElement = _dataValues[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
    }

    return marshalSize;
}

